rfcs/text/0430-hook-service.md

```diff
@@ -0,0 +1,148 @@
+- Feature Name: `hook-service`
+- Start Date: (fill me in with today's date, 2022-07-12)
+- RFC PR: [blockprotocol/blockprotocol#430](https://github.com/blockprotocol/blockprotocol/pull/430)
+- Block Protocol Discussion: N/A
+
+# Summary
+
+[summary]: #summary
+
+This RFC introduces a Hook Service which standardizes the communication necessary between blocks and embedding applications, allowing the latter to 'take over' rendering one or more parts of a block. This allows functionality to be extended, while ensuring blocks can remain uncoupled from embedding applications.
+
+It also extends the Core Specification so that embedding applications must respond with a `NOT_IMPLEMENTED` error for messages to any Block Protocol services it does not implement. It is our stated intention to release an update to the `@blockprotocol/core` package which will do this for embedding applications.
+
+# Motivation
+
+[motivation]: #motivation
+
+It is not uncommon for embedding applications to want to provide a consistent 'look and feel' across distinct blocks which implement similar features – i.e, rich text or image editing. Additionally, the implementation of these features may require integrating with systems bespoke to the embedding application. We don't want blocks to have to 'know' about the implementation of this functionality, or necessarily the shape of the data involved in providing this functionality. Therefore, the embedding application needs to be responsible for providing this functionality, but this depends on cooperation by block authors.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Hook Service
+
+A **hook** is an injection point blocks provide to embedding applications to allow them to optionally render or modify already rendered views associated with a Graph Service property in the block's schema.
+
+**Note: the word 'optionally' here is key.** Blocks cannot today _require_ embedding applications to implement a handler for any specific 'hook'. Instead, they should implement a fallback in the event the embedding application does not implement a hook (see the Errors section below)
+
+A classical example would be rich text editing. In this case, the data structure for formatted text may be bespoke to the embedding application, and impossible for the block to predictably parse or generate. The expected text editing experience may involve custom formatting controls, or special inline actions or commands. By injecting its own rich text editing input, an embedding application can provide all this.
+
+A block implements a hook by sending a `hook` message to the embedding application, with five pieces of information:
```

Comment:
I think the existing content can serve as the reference-level explanation, with a bit more by way of real-world example to fill out the guide-level
```suggestion

In the rich text example, a block would send a message under the service specifying a type of `"text"`, a DOM node in which to inject the rich text editing input, and the property path the data should be stored under. Any data created by the input can be stored by the embedding application in its own format. The block need not receive this data, since it won't understand it anyway. Instead, the application should send the block a value for the property which matches what the block expects in its schema. In the case of rich text, this is likely to be `"string"`, and the application may choose to supply a plain text representation of the rich text created.

# Reference-level explanation

[reference-level-explanation]: #reference-level-explanation

A block implements a hook by sending a `hook` message to the embedding application, with five pieces of information:
```